### PR TITLE
Optimize MaxText unit and integration test suite runtime

### PR DESCRIPTION
**Pull request by shralex**
This pull request optimizes the MaxText TPU/GPU CI test suites to substantially reduce total execution duration. It eliminates redundant compilation, graph tracing, and setup latencies without affecting functional verification or reducing test coverage.

This PR saves ~35% (> 12m) for each CI test run (longest test time goes down from 34m+ to 22m). The following table details the total run time savings, which impact runner availability and queueing time:
Screenshot 2026-05-12 at 8 24 10 AM

Classes of Optimizations:

1. Model Downscaling (Compilation): Scaled down model configurations (embedding dimensions, heads, layers) across unit and integration tests to minimize XLA compilation overhead.
2. Turns out that if you have two decorators like this on a test:

@pytest.mark.cpu_only
@parameterized.named_parameters(
      {"testcase_name": "dot_product", "attention": "dot_product"},
      {"testcase_name": "tokamax_splash", "attention": "flash"},
  )
  def test_

Then then cpu_only marker has no effect and the test runs on TPUs. If you switch the order of the markers as this PR does, then the cpu_only marker works.

3. Sequence & Step Capping (Execution): Capped sequence lengths at 128 or 512 (down from 1024/8192) and reduced loop steps/step counts across test assertions and runs.
4. Cached Data & Model Pipelines (Initialization & Compilation):
Replaced method-level dataset recreation with class-level lazy caching to construct data pipelines once per suite. 
5. Moved input pipeline tests to be cpu_only.
6. Added class-level caching of unquantized base model results in QuantTest to eliminate redundant compilation and forward/backward pass execution across multiple quantization tests.
7. Local Path & Tokenizer Redirection (I/O & Network):
- Enforced synthetic datasets and redirected GCS asset paths to local mock directories to eliminate network/GCS latency.
- Redirected remote Hugging Face tokenizers (google-t5/t5-large, deepseek-ai/DeepSeek-V3, and zephyr-7b-beta) in HF/Grain tests to local pre-saved asset directories (tokenizer.default and qwen3-tokenizer), completely removing external network request overhead.
8. Leveraged JAX's standard __wrapped__ attribute to extract and trace the underlying un-decorated Python function. This bypasses the Pathways proxy compilation (which took minutes), performing client-side local CPU tracing. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by shralex** [comment]
@gemini-cli /investigate

**Review by igorts-git — APPROVED** [review]

File: src/maxtext/utils/maxtext_utils.py [hunk 1]
```diff
@@ -1936,8 +1936,19 @@ def maybe_dump_jaxpr(config, p_train_step, train_step_inputs):
...
+
+  def to_abstract(x):
+    if hasattr(x, "shape") and hasattr(x, "dtype"):
+      return jax.ShapeDtypeStruct(shape=x.shape, dtype=x.dtype)
+    return x
+
+  # Convert all input arguments recursively to purely local abstract ShapeDtypeStruct objects
+  # to completely bypass remote Array objects and proxy tracing overhead.
+  abstract_inputs = jax.tree.map(to_abstract, train_step_inputs)
+  p_train_jaxpr = jax.make_jaxpr(unwrapped_step)(*abstract_inputs)
```
Comment (NuojCheng) [inline src/maxtext/utils/maxtext_utils.py:1949]: do we observe shorter test time, e.g. for Jaxpr dump test in tpu integration test? If so, we should make similar change to train_compile.py cc @gobbleturk

**Review by NuojCheng — APPROVED** [review]

File: tests/integration/deepseek_scan_engram_test.py [hunk 2]
```diff
@@ -154,15 +173,16 @@ def test_decoder_init_engram_2_8(self, mock_from_pretrained):
     """Test engram layers at indices 2 and 8."""
```
Comment (RissyRan) [inline tests/integration/deepseek_scan_engram_test.py:173]: Could you help update the comments? Similar for other 2 tests. Thanks!

**Review by RissyRan — APPROVED** [review]
Thanks Alex! Minor comments.
